Service/Microsoft.Health.DeIdentification.Local/InMemoryQueueClient.cs

```diff
@@ -1,37 +1,106 @@
 ﻿using Microsoft.Health.JobManagement;
+using System.Collections.ObjectModel;
 
 namespace Microsoft.Health.DeIdentification.Local
 {
     public class InMemoryQueueClient : IQueueClient
     {
+        private int largestId = 1;
+        private List<JobInfo> jobInfos = new List<JobInfo>();
+        public List<JobInfo> JobInfos
+        {
+            get { return jobInfos; } 
+        }
+
         public Task CancelJobByGroupIdAsync(byte queueType, long groupId, CancellationToken cancellationToken)
         {
             throw new NotImplementedException();
         }
 
         public Task CancelJobByIdAsync(byte queueType, long jobId, CancellationToken cancellationToken)
         {
-            throw new NotImplementedException();
+            foreach (JobInfo jobInfo in jobInfos.Where(t => t.Id == jobId))
+            {
+                if (jobInfo.Status == JobStatus.Created)
+                {
+                    jobInfo.Status = JobStatus.Cancelled;
+                }
+
+                if (jobInfo.Status == JobStatus.Running)
+                {
+                    jobInfo.CancelRequested = true;
+                }
+            }
+
+            return Task.CompletedTask;
         }
 
-        public Task CompleteJobAsync(JobInfo jobInfo, bool requestCancellationOnFailure, CancellationToken cancellationToken)
+        public async Task CompleteJobAsync(JobInfo jobInfo, bool requestCancellationOnFailure, CancellationToken cancellationToken)
         {
-            throw new NotImplementedException();
+            JobInfo jobInfoStore = jobInfos.FirstOrDefault(t => t.Id == jobInfo.Id);
+            jobInfoStore.Status = jobInfo.Status;
+            jobInfoStore.Result = jobInfo.Result;
+
+            if (requestCancellationOnFailure && jobInfo.Status == JobStatus.Failed)
+            {
+                await CancelJobByIdAsync(jobInfo.QueueType, jobInfo.Id, cancellationToken);
+            }
         }
 
         public Task<JobInfo> DequeueAsync(byte queueType, string worker, int heartbeatTimeoutSec, CancellationToken cancellationToken, long? jobId = null)
         {
-            throw new NotImplementedException();
+            JobInfo job = jobInfos.FirstOrDefault(t => t.Status == JobStatus.Created || (t.Status == JobStatus.Running && (DateTime.Now - t.HeartbeatDateTime) > TimeSpan.FromSeconds(heartbeatTimeoutSec)));
```

Comment:
what's the order of job dequeuing? is it possible that there is always a new job adding, and dequeue the new job, so the old job will not be dequeued?

Reply:
I will verify the order in test